## Basic Python - Project <a id='intro'></a>

## Introduction <a id='intro'></a>
In this project, you will work with data from the entertainment industry. You will study a dataset with records on movies and shows. The research will focus on the "Golden Age" of television, which began in 1999 with the release of *The Sopranos* and is still ongoing.

The aim of this project is to investigate how the number of votes a title receives impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the "Golden Age" of television also have the most votes.

### Stages 
Data on movies and shows is stored in the `/datasets/movies_and_shows.csv` file. There is no information about the quality of the data, so you will need to explore it before doing the analysis.

First, you'll evaluate the quality of the data and see whether its issues are significant. Then, during data preprocessing, you will try to account for the most critical problems.
 
Your project will consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Data analysis

## Stage 1. Data overview <a id='data_review'></a>

Open and explore the data.

You'll need `pandas`, so import it.

In [1]:
# importing pandas
import pandas as pd


Read the `movies_and_shows.csv` file from the `datasets` folder and save it in the `df` variable:

In [2]:
# reading the files and storing them to df
try:
    df = pd.read_csv("movies_and_shows.csv")
except:
    df = pd.read_csv("/datasets/movies_and_shows.csv")

Print the first 10 table rows:

In [3]:
# obtaining the first 10 rows from the df table
# hint: you can use head() and tail() in Jupyter Notebook without wrapping them into print()
first_10_rows = df.head(10)
first_10_rows

name                      Character   r0le        TITLE   Type  \
0   Robert De Niro                  Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster                  Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                            Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                          Betsy  ACTOR  Taxi Driver  MOVIE   
5      Peter Boyle                         Wizard  ACTOR  Taxi Driver  MOVIE   
6   Leonard Harris      Senator Charles Palantine  ACTOR  Taxi Driver  MOVIE   
7   Diahnne Abbott                Concession Girl  ACTOR  Taxi Driver  MOVIE   
8      Gino Ardito             Policeman at Rally  ACTOR  Taxi Driver  MOVIE   
9  Martin Scorsese  Passenger Watching Silhouette  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0  
5          1976  ['drama', 'crime']         8.2    808582.0  
6          1976  ['drama', 'crime']         8.2    808582.0  
7          1976  ['drama', 'crime']         8.2    808582.0  
8          1976  ['drama', 'crime']         8.2    808582.0  
9          1976  ['drama', 'crime']         8.2    808582.0

Obtain the general information about the table with one command:

In [4]:
# obtaining general information about the data in df
df.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB


The table contains nine columns. The majority store the same data type: object. The only exceptions are `'release Year'` (int64 type), `'imdb sc0re'` (float64 type) and `'imdb v0tes'` (float64 type). Scores and votes will be used in our analysis, so it's important to verify that they are present in the dataframe in the appropriate numeric format. Three columns (`'TITLE'`, `'imdb sc0re'` and `'imdb v0tes'`) have missing values.

According to the documentation:
- `'name'` — actor/director's name and last name
- `'Character'` — character played (for actors)
- `'r0le '` — the person's contribution to the title (it can be in the capacity of either actor or director)
- `'TITLE '` — title of the movie (show)
- `'  Type'` — show or movie
- `'release Year'` — year when movie (show) was released
- `'genres'` — list of genres under which the movie (show) falls
- `'imdb sc0re'` — score on IMDb
- `'imdb v0tes'` — votes on IMDb

We can see three issues with the column names:
1. Some names are uppercase, while others are lowercase.
2. There are names containing whitespace.
3. A few column names have digit '0' instead of letter 'o'. 


### Conclusions <a id='data_review_conclusions'></a> 

Each row in the table stores data about a movie or show. The columns can be divided into two categories: the first is about the roles held by different people who worked on the movie or show (role, name of the actor or director, and character if the row is about an actor); the second category is information about the movie or show itself (title, release year, genre, imdb figures).

It's clear that there is sufficient data to do the analysis and evaluate our assumption. However, to move forward, we need to preprocess the data.

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>
Correct the formatting in the column headers and deal with the missing values. Then, check whether there are duplicates in the data.

In [5]:
# the list of column names in the df table

columns = df.columns.tolist()




Change the column names according to the rules of good style:
* If the name has several words, use snake_case
* All characters must be lowercase
* Remove whitespace
* Replace zero with letter 'o'

In [6]:
# renaming columns

# Convert column names to lowercase
df.columns = df.columns.str.lower()

# Remove teh spaces 
df.columns = df.columns.str.strip()

# Replace '0' with 'o'
df.columns = df.columns.str.replace('0', 'o')

# Replace spaces with underscores (snake_case)
df.columns = df.columns.str.replace(' ', '_')





Check the result. Print the names of the columns once more:

In [7]:
# checking result: the list of column names

# Print the modified column names
print(df.columns.tolist())

['name', 'character', 'role', 'title', 'type', 'release_year', 'genres', 'imdb_score', 'imdb_votes']


### Missing values <a id='missing_values'></a>
First, find the number of missing values in the table. To do so, combine two `pandas` methods:

In [8]:
# calculating missing values

# Count the number of missing values in each column
missing_values_count = df.isnull().sum()

# Print the result
print(missing_values_count)

name               0
character          0
role               0
title              1
type               0
release_year       0
genres             0
imdb_score      4609
imdb_votes      4726
dtype: int64


Not all missing values affect the research: the single missing value in `'title'` is not critical. The missing values in columns `'imdb_score'` and `'imdb_votes'` represent around 6% of all records (4,609 and 4,726, respectively, of the total 85,579). This could potentially affect our research. To avoid this issue, we will drop rows with missing values in the `'imdb_score'` and `'imdb_votes'` columns.

In [9]:
# dropping rows where columns with title, scores and votes have missing values

# Dropping rows with missing values in 'title', 'imdb_score', and 'imdb_votes' columns
df.dropna(subset=['title', 'imdb_score', 'imdb_votes'], how='any', inplace=True)

# Confirming the removal of rows
print("Number of rows after dropping missing values:", len(df))

Number of rows after dropping missing values: 80853


Make sure the table doesn't contain any more missing values. Count the missing values again.

In [10]:
# counting missing values

# Count the number of missing values in each column
missing_values_count = df.isnull().sum()

# Print the result
print(missing_values_count)

name            0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imdb_score      0
imdb_votes      0
dtype: int64


### Duplicates <a id='duplicates'></a>
Find the number of duplicate rows in the table using one command:

In [11]:
# counting duplicate rows

# Count the number of duplicate rows
duplicate_rows_count = df.duplicated().sum()

# Print the result
print("Number of duplicate rows:", duplicate_rows_count)


Number of duplicate rows: 6994


Review the duplicate rows to determine if removing them would distort our dataset.

In [12]:
# Produce table with duplicates (with original rows included) and review last 5 rows

# Select duplicate rows
duplicate_rows = df[df.duplicated()]

# Review the last 5 rows of the duplicate rows DataFrame
print(duplicate_rows.sample(10))


                     name         character   role  \
80688        Daniel Sauli   Officer Quinlan  ACTOR   
78981      Olga Bo?Ž?d§              Gabi  ACTOR   
84539     Angelica Giusto          Cristina  ACTOR   
84085      Serta?? G?¬der           unknown  ACTOR   
78141          John Mayer              Self  ACTOR   
84986       Vincent Leong   Old Chinese Man  ACTOR   
81925          Fa Zhiyuan            Yu She  ACTOR   
77813       Laura Stracko          Ensemble  ACTOR   
79951  Gianluca De Marchi  Domenico Moretti  ACTOR   
82630      Mustapha Aarab         John John  ACTOR   

                                    title   type  release_year  \
80688                   This Is the Night  MOVIE          2021   
78981                    The Getaway King  MOVIE          2021   
84539              Still Out of My League  MOVIE          2021   
84085                       In Good Hands  MOVIE          2022   
78141  Dirty Daddy: The Bob Saget Tribute  MOVIE          2022   
84986    

There are two clear duplicates in the printed rows. We can safely remove them.
Call the `pandas` method for getting rid of duplicate rows:

In [13]:
# removing duplicate rows

# Remove duplicate rows from the DataFrame
df.drop_duplicates(inplace=True)


Check for duplicate rows once more to make sure you have removed all of them:

In [14]:
# checking for duplicates

# Count the number of duplicate rows
duplicate_rows_count_after_removal = df.duplicated().sum()

# Print the result
print("Number of duplicate rows after removal:", duplicate_rows_count_after_removal)

Number of duplicate rows after removal: 0


Now get rid of implicit duplicates in the `'type'` column. For example, the string `'SHOW'` can be written in different ways. These kinds of errors will also affect the result.

Print a list of unique `'type'` names, sorted in alphabetical order. To do so:
* Retrieve the intended dataframe column 
* Apply a sorting method to it
* For the sorted column, call the method that will return all unique column values

In [15]:
# viewing unique type names

# Retrieve the 'type' column and remove leading/trailing whitespaces

df['type'] = df['type'].str.strip()

# Sort the 'type' column alphabetically
sorted_types = df['type'].sort_values()

# Get unique type names
unique_types = sorted_types.unique()

# Print the list of unique type names
print("Unique 'type' names:")
for type_name in unique_types:
    print(type_name)


Unique 'type' names:
MOVIE
SHOW
movies
shows
the movie
tv
tv series
tv show
tv shows


Look through the list to find implicit duplicates of `'show'` (`'movie'` duplicates will be ignored since the assumption is about shows). These could be names written incorrectly or alternative names of the same genre.

You will see the following implicit duplicates:
* `'shows'`
* `'SHOW'`
* `'tv show'`
* `'tv shows'`
* `'tv series'`
* `'tv'`

To get rid of them, declare the function `replace_wrong_show()` with two parameters: 
* `wrong_shows_list=` — the list of duplicates
* `correct_show=` — the string with the correct value

The function should correct the names in the `'type'` column from the `df` table (i.e., replace each value from the `wrong_shows_list` list with the value in `correct_show`).

In [16]:
# function for replacing implicit duplicates

def replace_wrong_show(df, wrong_shows_list, correct_show):
    for wrong_show in wrong_shows_list:
        df['type'] = df['type'].replace(wrong_show, correct_show)

# Example usage:
wrong_shows_list = ['shows', 'SHOW', 'tv show', 'tv shows', 'tv series', 'tv']
correct_show = 'TV Show'
replace_wrong_show(df, wrong_shows_list, correct_show)


Call `replace_wrong_show()` and pass it arguments so that it clears implicit duplicates and replaces them with `SHOW`:

In [17]:
# removing implicit duplicates

# def replace_wrong_show(df, wrong_shows_list, correct_show):
#     for wrong_show in wrong_shows_list:
#         df['type'] = df['type'].replace(wrong_show, correct_show)

# # Define the list of wrong show names and the correct show name
# wrong_shows_list = ['shows', 'SHOW', 'tv show', 'tv shows', 'tv series', 'tv']
# correct_show = 'SHOW'

# # Call the function to replace implicit duplicates
# replace_wrong_show(df, wrong_shows_list, correct_show)


Make sure the duplicate names are removed. Print the list of unique values from the `'type'` column:

In [18]:
# viewing unique genre names


# Get unique values from the 'type' column
unique_types_after_correction = df['type'].unique()

# Print the list of unique values
print("Unique 'type' names after correction:")
for type_name in unique_types_after_correction:
    print(type_name)


Unique 'type' names after correction:
MOVIE
the movie
TV Show
movies


### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected three issues with the data:

- Incorrect header styles
- Missing values
- Duplicate rows and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All rows with missing values have been removed. 

The absence of duplicates will make the results more precise and easier to understand.

Now we can move on to our analysis of the prepared data.

## Stage 3. Data analysis <a id='hypotheses'></a>

Based on the previous project stages, you can now define how the assumption will be checked. Calculate the average amount of votes for each score (this data is available in the `imdb_score` and `imdb_votes` columns), and then check how these averages relate to each other. If the averages for shows with the highest scores are bigger than those for shows with lower scores, the assumption appears to be true.

Based on this, complete the following steps:

- Filter the dataframe to only include shows released in 1999 or later.
- Group scores into buckets by rounding the values of the appropriate column (a set of 1-10 integers will help us make the outcome of our calculations more evident without damaging the quality of our research).
- Identify outliers among scores based on their number of votes, and exclude scores with few votes.
- Calculate the average votes for each score and check whether the assumption matches the results.

To filter the dataframe and only include shows released in 1999 or later, you will take two steps. First, keep only titles published in 1999 or later in our dataframe. Then, filter the table to only contain shows (movies will be removed).

In [19]:
# using conditional indexing modify df so it has only titles released after 1999 (with 1999 included)

df_after_1999 = df[df['release_year'] >= 1999]

# give the slice of dataframe new name

df_after_1999_name = df_after_1999.copy() 


In [20]:
# repeat conditional indexing so df has only shows (movies are removed as result)

df_shows_only = df_after_1999[df_after_1999['type'] == 'TV Show']

# Rename the sliced DataFrame for clarity
df_shows_only_name = df_shows_only.copy()


The scores that are to be grouped should be rounded. For instance, titles with scores like 7.8, 8.1, and 8.3 will all be placed in the same bucket with a score of 8.

In [21]:
# rounding column with scores

# Round the 'imdb_score' column to the nearest integer
df_shows_only['rounded_score'] = df_shows_only['imdb_score'].round().astype(int)

#checking the outcome with tail()

print(df_shows_only.tail())


                      name character      role      title     type  \
85433     Maneerat Kam-Uan        Ae     ACTOR  Let's Eat  TV Show   
85434    Rudklao Amratisha   unknown     ACTOR  Let's Eat  TV Show   
85435      Jaturong Mokjok   unknown     ACTOR  Let's Eat  TV Show   
85436     Pisamai Wilaisak   unknown     ACTOR  Let's Eat  TV Show   
85437  Sarawut Wichiensarn   unknown  DIRECTOR  Let's Eat  TV Show   

       release_year               genres  imdb_score  imdb_votes  \
85433          2021  ['drama', 'comedy']         8.2         5.0   
85434          2021  ['drama', 'comedy']         8.2         5.0   
85435          2021  ['drama', 'comedy']         8.2         5.0   
85436          2021  ['drama', 'comedy']         8.2         5.0   
85437          2021  ['drama', 'comedy']         8.2         5.0   

       rounded_score  
85433              8  
85434              8  
85435              8  
85436              8  
85437              8  


/tmp/ipykernel_50/2903679242.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_shows_only['rounded_score'] = df_shows_only['imdb_score'].round().astype(int)


It is now time to identify outliers based on the number of votes.

In [23]:
# Use groupby() for scores and count all unique values in each group, print the result

score_counts = df_shows_only.groupby("rounded_score")['title'].nunique()

# Print the result
print(score_counts)


rounded_score
2       5
3       4
4      30
5      94
6     347
7     593
8     539
9      65
10      1
Name: title, dtype: int64


Based on the aggregation performed, it is evident that scores 2 (24 voted shows), 3 (27 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

To obtain the mean numbers of votes for the selected scores (we identified a range of 4-9 as acceptable), use conditional filtering and grouping.

In [24]:
# filter dataframe using two conditions (scores to be in the range 4-9)

# group scores and corresponding average number of votes, reset index and print the result


# Filter the DataFrame to include scores in the range 4-9
filtered_df = df_shows_only[df_shows_only['rounded_score'].between(4, 9)]

# Group scores and calculate the mean number of votes
mean_votes_per_score = filtered_df.groupby('rounded_score')['imdb_votes'].mean().reset_index()

# Print the result
print(mean_votes_per_score)


   rounded_score     imdb_votes
0              4    5277.583333
1              5    3143.942568
2              6    3481.717322
3              7    8727.068211
4              8   30299.460967
5              9  126904.109515


Now for the final step! Round the column with the averages, rename both columns, and print the dataframe in descending order.

In [25]:
# round column with averages

# rename columns

# print dataframe in descending order


# Round the 'imdb_votes' column in the DataFrame mean_votes_per_score
mean_votes_per_score['mean_votes'] = mean_votes_per_score['imdb_votes'].round().astype(int)

# Drop the original 'imdb_votes' column
mean_votes_per_score.drop(columns=['imdb_votes'], inplace=True)

# Rename the columns
mean_votes_per_score.rename(columns={'rounded_score': 'Rounded Score', 'mean_votes': 'Mean Votes'}, inplace=True)

# Print the DataFrame in descending order
print(mean_votes_per_score.sort_values(by='Mean Votes', ascending=False))

   Rounded Score  Mean Votes
5              9      126904
4              8       30299
3              7        8727
0              4        5278
2              6        3482
1              5        3144


The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

## Conclusion <a id='hypotheses'></a>

The research done confirms that highly-rated shows released during the "Golden Age" of television also have the most votes. While shows with score 4 have more votes than ones with scores 5 and 6, the top three (scores 7-9) have the largest number. The data studied represents around 94% of the original set, so we can be confident in our findings.